src/devices/Si7021/Si7021.cs

```diff
@@ -3,59 +3,192 @@
+        /// <param name="resolution">Si7021 Read Resolution</param>
+        public Si7021(I2cDevice i2cDevice, Resolution resolution = Resolution.Resolution1)
         {
-            double tempCelcius = ReadTemperatureInCelcius();
-            double tempFahrenheit = (tempCelcius * (9 / 5)) + 32;
-            
-            return tempFahrenheit;
+            _i2cDevice = i2cDevice;
+
+            SetResolution(resolution);
         }
-        
-        public double ReadTemperatureInCelcius()
+
+        /// <summary>
+        /// Get Si7021 Temperature (℃)
+        /// </summary>
+        /// <returns>Temperature (℃)</returns>
+        private double GetTemperature()
         {
-            byte[] buffer = new byte[2];
+            Span<byte> readbuff = stackalloc byte[2];
 
             // Send temperature command, read back two bytes
-            _i2cDevice.WriteByte(ReadTemperatureCommand);
-            _i2cDevice.Read(buffer.AsSpan());
-            
+            _i2cDevice.WriteByte((byte)Register.SI_TEMP);
+            // wait SCL free
+            Thread.Sleep(20);
+            _i2cDevice.Read(readbuff);
+
             // Calculate temperature
-            double temp_code = buffer[0] << 8 | buffer[1];
-            double tempCelcius = (((175.72 * temp_code) / 65536) - 46.85);
-            
-            return tempCelcius;
+            ushort raw = BinaryPrimitives.ReadUInt16BigEndian(readbuff);
+            double temp = 175.72 * raw / 65536.0 - 46.85;
+
+            return Math.Round(temp, 1);
         }
-        
-        public double ReadHumidity()
+
+        /// <summary>
+        /// Get Si7021 Relative Humidity (%)
+        /// </summary>
+        /// <returns>Relative Humidity (%)</returns>
+        private double GetHumidity()
         {
-            byte[] buffer = new byte[2];
-            
+            Span<byte> readbuff = stackalloc byte[2];
+
             // Send humidity read command, read back two bytes
-            _i2cDevice.WriteByte(ReadHumidityCommand);
-            _i2cDevice.Read(buffer.AsSpan());
-            
+            _i2cDevice.WriteByte((byte)Register.SI_HUMI);
+            // wait SCL free
+            Thread.Sleep(20);
+            _i2cDevice.Read(readbuff);
+
             // Calculate humidity
-            double rh_code = buffer[0] << 8 | buffer[1];
-            double humidity = ((125 * rh_code) / 65536) - 6;
-            
-            return humidity;
+            ushort raw = BinaryPrimitives.ReadUInt16BigEndian(readbuff);
+            double humidity = 125 * raw / 65536.0 - 6;
+
+            return Math.Round(humidity);
+        }
+
+        /// <summary>
+        /// Get Si7021 Firmware Revision
+        /// </summary>
+        /// <returns>Firmware Revision</returns>
+        private byte GetRevision()
+        {
+            Span<byte> writeBuff = stackalloc byte[2] { (byte)Register.SI_REVISION_MSB, (byte)Register.SI_REVISION_LSB };
```

Comment:
slightly confused - you are asking for 2 bytes but reading 1?

Reply:
Yes, you can see the `5.4. Firmware Revision` in the datasheet. Actually, this should be called command.😄